pkg/controller/workload/bpfcache/locality_cache.go

```diff
@@ -0,0 +1,141 @@
+	mask      uint32 // mask
+}
+
+func Valid(s string) bool {
+	return s != ""
+}
+
+func (l *localityInfo) Set(s string, param uint32) {
+	if !Valid(s) {
+		return
+	}
+	switch param {
+	case REGION:
+		l.region = s
+	case ZONE:
+		l.zone = s
+	case SUBZONE:
+		l.subZone = s
+	case NODENAME:
+		l.nodeName = s
+	case CLUSTERID:
+		l.clusterId = s
+	case NETWORK:
+		l.network = s
+	}
+	l.mask |= param
+}
+
+func (l *localityInfo) Clear(param uint32) {
+	l.mask &= ^param
+}
+
+func (l *localityInfo) IsSet(param uint32) bool {
+	return l.mask&param != 0
+}
+
+type LocalityCache struct {
+	mutex        sync.RWMutex
+	LocalityInfo *localityInfo
+}
+
+func NewLocalityCache() LocalityCache {
+	return LocalityCache{
+		LocalityInfo: nil,
+	}
+}
+
+func (l *LocalityCache) SetLocality(nodeName, clusterId, network string, locality *workloadapi.Locality) {
+	l.mutex.Lock()
+	defer l.mutex.Unlock()
+	if l.LocalityInfo == nil {
+		l.LocalityInfo = &localityInfo{}
+	}
+
+	// notice: nodeName should set by processor or os.Getenv("NODE_NAME"),
+	l.LocalityInfo.Set(nodeName, NODENAME)
+	l.LocalityInfo.Set(locality.GetRegion(), REGION)
+	l.LocalityInfo.Set(locality.GetSubzone(), SUBZONE)
+	l.LocalityInfo.Set(locality.GetZone(), ZONE)
+	l.LocalityInfo.Set(clusterId, CLUSTERID)
+	l.LocalityInfo.Set(network, NETWORK)
+}
+
+func (l *LocalityCache) CalcLocalityLBPrio(wl *workloadapi.Workload, rp []workloadapi.LoadBalancing_Scope) uint32 {
+	var rank uint32 = 0
+	for _, scope := range rp {
+		switch scope {
+		case workloadapi.LoadBalancing_REGION:
+			log.Debugf("l.LocalityInfo.region %#v, wl.GetLocality().GetRegion() %#v", l.LocalityInfo.region, wl.GetLocality().GetRegion())
+			if l.LocalityInfo.region == wl.GetLocality().GetRegion() {
+				rank++
+			}
+		case workloadapi.LoadBalancing_ZONE:
+			log.Debugf("l.LocalityInfo.zone %#v, wl.GetLocality().GetZone() %#v", l.LocalityInfo.zone, wl.GetLocality().GetZone())
+			if l.LocalityInfo.zone == wl.GetLocality().GetZone() {
+				rank++
+			}
+		case workloadapi.LoadBalancing_SUBZONE:
+			log.Debugf("l.LocalityInfo.subZone %#v, wl.GetLocality().GetSubzone() %#v", l.LocalityInfo.subZone, wl.GetLocality().GetSubzone())
+			if l.LocalityInfo.subZone == wl.GetLocality().GetSubzone() {
+				rank++
+			}
+		case workloadapi.LoadBalancing_NODE:
+			log.Debugf("l.LocalityInfo.nodeName %#v, wl.GetNode() %#v", l.LocalityInfo.nodeName, wl.GetNode())
+			if l.LocalityInfo.nodeName == wl.GetNode() {
+				rank++
+			}
+		case workloadapi.LoadBalancing_NETWORK:
+			log.Debugf("l.LocalityInfo.network %#v, wl.GetNetwork() %#v", l.LocalityInfo.network, wl.GetNetwork())
+			if l.LocalityInfo.network == wl.GetNetwork() {
+				rank++
+			}
+		case workloadapi.LoadBalancing_CLUSTER:
+			log.Debugf("l.LocalityInfo.clusterId %#v, wl.GetClusterId() %#v", l.LocalityInfo.clusterId, wl.GetClusterId())
+			if l.LocalityInfo.clusterId == wl.GetClusterId() {
+				rank++
+			}
+		}
+	}
+	return uint32(len(rp)) - rank
```

Comment:
This method is not right. Keep in mind, we shoud match rp []workloadapi.LoadBalancing_Scope in order, if one doesnot match, we should stop.   


So i am super sure this result now is nort right.

Reply:
fixed it

Reply:
There is not test coverage, i will write a test later to test your function.  I am feeling this function is still not correct